Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -305,31 +305,63 @@ public async Task<string> WalkTree(string treeNodeKey)
-                        this.Parameters.UserContext,
-                        this.Parameters.TreeName,
-                        this.Parameters.RootSessionId,
-                        this.walkTreeCts.Token).ConfigureAwait(false);
+                    TreeNodeContext treeNodeContext = null;
+                    if (this.Parameters.CallbacksV2 != null)
+                    {
+                        // If applicable, use the new CallbacksV2 with ITreeWalkerCallbacksV2
+                        // Evaluate the dynamic properties that are used by the actionTask.
+                        treeNodeContext = new TreeNodeContext(
+                            this.Parameters.SessionId,
+                            treeNodeKey,
+                            await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
+                            this.Parameters.UserContext,
+                            this.walkTreeCts.Token,
+                            this.Parameters.TreeName,
+                            this.Parameters.RootSessionId,
+                            shouldSkipActionsInTreeNode:false
+                        );
+
+                        // Follow the previous Callbacks with ITreeWalkerCallbacks.
+                        await this.Parameters.CallbacksV2.BeforeVisitNode(treeNodeContext).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        // Follow the previous Callbacks with ITreeWalkerCallbacks.
+                        await this.Parameters.Callbacks.BeforeVisitNode(
+                            this.Parameters.SessionId,
+                            current,
+                            await this.EvaluateDynamicProperty(this.Schema.Tree[current].Properties, null),
+                            this.Parameters.UserContext,
+                            this.Parameters.TreeName,
+                            this.Parameters.RootSessionId,
+                            this.walkTreeCts.Token).ConfigureAwait(false);
+                    }
 
                     try
                     {
                         // Exceptions are thrown here if VisitNode hit a timeout, was cancelled, or failed.
-                        next = await this.VisitNode(current).ConfigureAwait(false);
+                        bool skipAction = treeNodeContext != null && treeNodeContext.ShouldSkipActionsInTreeNode;
+                        next = await this.VisitNode(current, skipAction).ConfigureAwait(false);
                     }
                     finally
                     {
-                        // Always call this callback, whether or not visit node was successful.
-                        await this.Parameters.Callbacks.AfterVisitNode(
+                        if (treeNodeContext != null)
+                        {
+                            // Follow the previous Callbacks with ITreeWalkerCallbacks.
```

Comment:
>// Follow the previous Callbacks with ITreeWalkerCallbacks. [](start = 28, length = 59)

Comment does not line up with V2 scenario.

Reply:
fixed the comment.

---
In reply to: [504165345](https://github.com/microsoft/Forge/pull/32#discussion_r504165345) [](ancestors = 504165345)